common/metrics/opentelemetry_scope.go

```diff
@@ -80,12 +81,18 @@ func (m *opentelemetryScope) IncCounter(id int) {
 func (m *opentelemetryScope) AddCounter(id int, delta int64) {
 	def := m.defs[id]
 	ctx := context.Background()
-	m.meterMust.NewInt64Counter(def.metricName.String()).Add(ctx, delta, m.labels...)
+	c, err := m.meter.SyncInt64().Counter(def.metricName.String())
+	if err != nil {
+		panic(err)
+	}
+	c.Add(ctx, delta, m.labels...)
 
 	if !def.metricRollupName.Empty() && (m.rootScope != nil) {
-		m.rootScope.meterMust.NewInt64Counter(def.metricRollupName.String()).Add(
-			ctx, delta, m.rootScope.labels...,
-		)
+		c, err := m.rootScope.meter.SyncInt64().Counter(def.metricRollupName.String())
```

Comment:
I wonder if we could support sync/async by using the InstrumentProvider interface rather than deciding at instantiation here. I think I am going to dig further into these interfaces as time goes on.


https://github.com/open-telemetry/opentelemetry-go/blob/main/metric/instrument/syncint64/syncint64.go#L25
and
https://github.com/open-telemetry/opentelemetry-go/blob/main/metric/instrument/asyncint64/asyncint64.go#L25

Reply:
👍  Looks like only `gaugeCache` is using async and it is created outside the scope implementation. So looks promising we can switch to using InstrumentProvider here.